dev/com.ibm.ws.microprofile.graphql.1.0/src/com/ibm/ws/microprofile/graphql/component/AuthInterceptor.java

```diff
@@ -0,0 +1,123 @@
+import javax.interceptor.AroundInvoke;
+import javax.interceptor.Interceptor;
+import javax.interceptor.InvocationContext;
+import javax.servlet.http.HttpServletRequest;
+
+import org.eclipse.microprofile.graphql.GraphQLApi;
+
+import com.ibm.websphere.ras.Tr;
+import com.ibm.websphere.ras.TraceComponent;
+
+@Dependent
+@GraphQLApi
+@Interceptor
+@Priority(Interceptor.Priority.PLATFORM_BEFORE)
+public class AuthInterceptor {
+
+    private final static TraceComponent tc = Tr.register(AuthInterceptor.class);
+
+    @AroundInvoke
+    public Object checkAuthorized(InvocationContext ctx) throws Exception {
+        Method m = ctx.getMethod();
+
+        if (isAuthorized(m)) {
+            return ctx.proceed();
+        }
+        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+            Tr.debug(tc, "Unauthorized!");
+        }
+        // TODO: determine if this is the right thing to do here... maybe different exception?
+        throw new Exception("Unauthorized");
+    }
+
+    private static boolean isAuthorized(Method m) {
+        // First check annotations on methods
+        if (m.getAnnotation(DenyAll.class) != null) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(tc, "@DenyAll on method");
+            }
+            return false;
+        }
+        if (m.getAnnotation(PermitAll.class) != null) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(tc, "@PermitAll on method");
+            }
+            return true;
+        }
+        RolesAllowed rolesAllowed = m.getAnnotation(RolesAllowed.class);
+        if (rolesAllowed != null) {
+            HttpServletRequest req = AuthFilter.getCurrentRequest();
+            for (String role : rolesAllowed.value()) {
+                if (req.isUserInRole(role)) {
+                    if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                        Tr.debug(tc, "User " + req.getUserPrincipal() + " is in (method) @RolesAllowed declared role, " + role);
+                    }
+                    return true;
+                }
+            }
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(tc, "User " + req.getUserPrincipal() + " is not in role defined in @RolesAllowed on method.");
+            }
+            return false;
+        }
+
+        // if method has no annotations, check class level
+        Class<?> cls = m.getDeclaringClass();
+        if (cls.getAnnotation(DenyAll.class) != null) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(tc, "@DenyAll on class");
+            }
+            return false;
+        }
+        if (cls.getAnnotation(PermitAll.class) != null) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(tc, "@PermitAll on class");
```

Comment:
similar to above, do we want to trace the class name?

Reply:
For the class name, we trace that already in https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.ws.microprofile.graphql.1.0/src/com/ibm/ws/microprofile/graphql/component/TracingInterceptor.java